packages/astro/src/internal/custom-elements/protect.ts

```diff
@@ -0,0 +1,64 @@
+import type { CheckAuthorizationWithCustomPermissions } from '@clerk/types';
+
+import { $authStore } from '../../stores/external';
+
+export class Protect extends HTMLElement {
+  private defaultSlot: HTMLDivElement | null = null;
+  private fallbackSlot: HTMLDivElement | null = null;
+  private authStoreListener: (() => void) | null = null;
+
+  constructor() {
+    super();
+  }
+
+  connectedCallback() {
+    this.toggleContentVisibility();
+  }
+
+  disconnectedCallback() {
+    this.authStoreListener?.();
+  }
+
+  toggleContentVisibility() {
+    this.defaultSlot = this.querySelector('[data-default-slot]');
+    this.fallbackSlot = this.querySelector('[data-fallback-slot]');
+
+    this.authStoreListener = $authStore.subscribe(state => {
+      const has = (params: Parameters<CheckAuthorizationWithCustomPermissions>[0]) => {
+        if (!params?.permission && !params?.role) {
+          throw new Error(
+            'Missing parameters. The prop permission or role is required to be passed. Example usage: `has({permission: "org:posts:edit"})`',
+          );
+        }
+
+        if (!state.orgId || !state.userId || !state.orgRole || !state?.orgPermissions) {
+          return false;
+        }
+
+        if (params.permission) {
+          return state.orgPermissions.includes(params.permission);
+        }
+
+        if (params.role) {
+          return state.orgRole === params.role;
+        }
+
+        return false;
+      };
```

Comment:
❓ Do we need to define `has` here ? Can we get access to `checkAuthorization` from `$session` ?

If not, please add a TODO comment for us in order to address this duplicate code issue. We are starting to redefine `has` in too many places.

Reply:
Interesting, that's cool. Yeah `checkAuthorization` works. So is it correct that we're doing `has` in the `useAuth` hooks just to throw a specific error?

Applied this change btw